### PR TITLE
Implement logResponseDiscrepencies functions

### LINKED ISSUE #9793: implement logResponseDiscrepencies functions
logResponseDiscrepencies is a helper function in [combiner/src/pnp/endpoints/sign/action.ts](https://github.com/celo-org/celo-monorepo/blob/5b5be63ac7be81873645c230aac3df5e6b468a1a/packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts#L43) that logs any differences in between the responses the combiner receives from each signer. 

- [ ] audit the existing logResponseDiscrepencies function, as this is mostly old code that could use another pass over. Make any improvements / simplifications that come to mind. 
- [ ] implement the empty logResponseDiscrepencies function in [combiner/src/domain/endpoints/sign/action.ts](url)
- [ ] add an analogous function for the domain/disable endpoint
- [ ] add an analogous function for the domain/quota endpoint
- [ ] add an analogous function for the pnp/quota endpoint
- [ ] Probably easiest to test visually by running the integration tests. We may need to add some test cases to do this thoroughly

### PR DESCRIPTION
**Pull request by isabellewei**
### Description

Implemented logging of discrepancies in combiner endpoints:
- pnp/sign
- domain/sign
- pnp/quota
- domain/quota
- skipped domain/disable because no meaningful params to check discrepancies for

### Tested

Need to add tests

### Related issues

- Fixes #9793

### CODE REVIEW

File: packages/phone-number-privacy/combiner/src/domain/endpoints/quota/action.ts [hunk 1]
```diff
@@ -31,4 +37,37 @@ export class DomainQuotaAction extends CombineAction<DomainQuotaStatusRequest> {
...
+          status: response.res.status,
+        }
+    )
+    statuses.forEach((status) => {
+      if (Math.abs(status.now - expectedStatus.now) > MAX_TIMESTAMP_DISCREPANCY_THRESHOLD) {
+        session.logger.warn({ values }, 'Inconsistent signer domain server timestamps')
+        return
+      }
+    })
+    session.logger.debug({ values }, 'Sequential Delay domain state')
```
Comment (isabellewei) [inline packages/phone-number-privacy/combiner/src/domain/endpoints/quota/action.ts]: always logging the sequential delay domain state if none of the above errors have been triggered, just to have the info available in the logs

File: packages/phone-number-privacy/combiner/src/domain/endpoints/sign/action.ts [hunk 2]
```diff
@@ -45,11 +46,35 @@ export class DomainSignAction extends SignAction<DomainRestrictedSignatureReques
...
+    const values = successes.map((response) => {
+      return {
+        signer: response.url,
+        status: response.res.status!,
+      }
+    })
+    successes.forEach((signerResponse) => {
+      const status = signerResponse.res.status!
+      if (Math.abs(status.now - expectedStatus.now) > MAX_TIMESTAMP_DISCREPANCY_THRESHOLD) {
+        session.logger.warn({ values }, 'Inconsistent signer domain server timestamps')
```
Comment (isabellewei) [inline packages/phone-number-privacy/combiner/src/domain/endpoints/sign/action.ts]: is this serious enough to create a WarningMessage for?

File: packages/phone-number-privacy/common/src/utils/constants.ts [hunk 3]
```diff
@@ -7,3 +7,4 @@ export const RETRY_COUNT = 5
 export const RETRY_DELAY_IN_MS = 100
 export const MAX_BLOCK_DISCREPANCY_THRESHOLD = 3
 export const KEY_VERSION_HEADER = 'odis-key-version' // headers must be all lower case
+export const MAX_TIMESTAMP_DISCREPANCY_THRESHOLD = 30
```
Comment (alecps) [inline packages/phone-number-privacy/common/src/utils/constants.ts]: should these be moved into the combiner?

File: packages/phone-number-privacy/combiner/src/domain/endpoints/quota/action.ts [hunk 4]
```diff
@@ -31,4 +37,37 @@ export class DomainQuotaAction extends CombineAction<DomainQuotaStatusRequest> {
       session.logger
     )
   }
+
+  protected logResponseDiscrepancies(session: Session<DomainQuotaStatusRequest>): void {
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/domain/endpoints/quota/action.ts]: Where is this getting called?

File: packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts [hunk 5]
```diff
@@ -43,58 +44,55 @@ export class PnpSignAction extends SignAction<SignMessageRequest> {
...
   protected logResponseDiscrepancies(session: CryptoSession<SignMessageRequest>): void {
-    // Only compare responses which have values for the quota fields
-    const successes = session.responses.filter(
-      (signerResponse) =>
-        signerResponse.res &&
-        signerResponse.res.performedQueryCount &&
-        signerResponse.res.totalQuota &&
-        signerResponse.res.blockNumber
-    )
+    // TODO: responses should all already be successes due to CombineAction receiveSuccess
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts]: Could you make this label TODO(2.0.0) and include a link to the ticket you created in the comment?

File: packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts [hunk 6]
```diff
@@ -43,58 +44,55 @@ export class PnpSignAction extends SignAction<SignMessageRequest> {
...
-          return {
-            signer: _signerResponse.url,
-            performedQueryCount: _signerResponse.res.performedQueryCount,
-            totalQuota: _signerResponse.res.totalQuota,
-          }
-        })
+        const blockValues = session.responses
+          .map(
+            (response) =>
+              response.res.success && {
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts]: I'm a little confused about this. What is the type of this array?

File: packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts [hunk 7]
```diff
@@ -43,58 +44,55 @@ export class PnpSignAction extends SignAction<SignMessageRequest> {
...
+            signer: response.url,
+            blockNumber: response.res.blockNumber,
+            totalQuota: response.res.totalQuota,
+            performedQueryCount: response.res.performedQueryCount,
+          }
+      )
+      .filter((val) => val)
 
-    if (successes.length === 0) {
+    if (responses.length === 0) {
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts]: Shouldn't we put this directly after we create the responses array?

File: packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts (hunk 6 above)
Comment (alecps) [reply packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts]: It looks like it's boolean?

**Review by alecps — CHANGES_REQUESTED** [review]

File: packages/phone-number-privacy/common/src/utils/constants.ts (hunk 3 above)
Comment (isabellewei) [reply packages/phone-number-privacy/common/src/utils/constants.ts]: do you think we should create a `constants` file in the combiner?

File: packages/phone-number-privacy/combiner/src/domain/endpoints/quota/action.ts [hunk 8]
```diff
@@ -3,18 +3,21 @@ import { CombineAction } from '../../../common/combine'
...
 import { Session } from '../../../common/session'
 import { OdisConfig } from '../../../config'
+import { DomainDiscrepanciesLogger } from '../../services/logDiscrepancies'
 import { DomainThresholdStateService } from '../../services/thresholdState'
 
 export class DomainQuotaAction extends CombineAction<DomainQuotaStatusRequest> {
   constructor(
     readonly config: OdisConfig,
     readonly thresholdStateService: DomainThresholdStateService<DomainQuotaStatusRequest>,
+    readonly discrepanyLogger: DomainDiscrepanciesLogger,
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/domain/endpoints/quota/action.ts]: nit: discrepancyLogger is missing a 'c'

File: packages/phone-number-privacy/common/src/utils/constants.ts (hunk 3 above)
Comment (alecps) [reply packages/phone-number-privacy/common/src/utils/constants.ts]: I think we can just add them to the config file in the combiner

File: packages/phone-number-privacy/combiner/src/server.ts [hunk 9]
```diff
@@ -62,6 +64,7 @@ export function startCombiner(config: CombinerConfig, kit?: ContractKit) {
     new PnpQuotaAction(
       config.phoneNumberPrivacy,
       pnpThresholdStateService,
+      new PnpDiscrepanciesLogger(),
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/server.ts]: Why are these only injected into the quota endpoints? Don't we need them in the signing endpoints too?

File: packages/phone-number-privacy/common/src/interfaces/errors.ts [hunk 10]
```diff
@@ -44,7 +44,6 @@ export enum WarningMessage {
   INCONSISTENT_SIGNER_DOMAIN_DISABLED_STATES = `CELO_ODIS_WARN_15 SIGNER Discrepency found in signer domain disabled states`,
   INVALID_AUTH_SIGNATURE = `CELO_ODIS_WARN_12 BAD_INPUT Authorization signature was incorrectly generated. Request will be rejected in a future version.`,
   INVALID_NONCE = `CELO_ODIS_WARN_13 BAD_INPUT SequentialDelayDomain nonce check failed on Signer request`,
-  INCONSISTENT_TOTAL_QUOTA = `CELO_ODIS_WARN_14 QUOTA Inconsistent totalQuota value from signers`,
```
Comment (alecps) [inline packages/phone-number-privacy/common/src/interfaces/errors.ts:47]: Why did we remove this?

File: packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts [hunk 11]
```diff
@@ -1,16 +1,18 @@
...
-  MAX_BLOCK_DISCREPANCY_THRESHOLD,
   SignMessageRequest,
   WarningMessage,
 } from '@celo/phone-number-privacy-common'
 import { CryptoSession } from '../../../common/crypto-session'
 import { SignAction } from '../../../common/sign'
+import { PnpDiscrepanciesLogger } from '../../services/logDiscrepancies'
 
 export class PnpSignAction extends SignAction<SignMessageRequest> {
+  readonly discrepanyLogger: PnpDiscrepanciesLogger = new PnpDiscrepanciesLogger()
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts]: nit: misspelling

File: packages/phone-number-privacy/combiner/src/pnp/endpoints/quota/action.ts [hunk 12]
```diff
@@ -3,26 +3,36 @@ import { CombineAction } from '../../../common/combine'
...
 import { Session } from '../../../common/session'
 import { OdisConfig } from '../../../config'
+import { PnpDiscrepanciesLogger } from '../../services/logDiscrepancies'
 import { PnpThresholdStateService } from '../../services/thresholdState'
 
 export class PnpQuotaAction extends CombineAction<PnpQuotaRequest> {
   constructor(
     readonly config: OdisConfig,
     readonly thresholdStateService: PnpThresholdStateService<PnpQuotaRequest>,
+    readonly discrepanyLogger: PnpDiscrepanciesLogger,
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/pnp/endpoints/quota/action.ts]: misspelling

File: packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts (hunk 11 above)
Comment (alecps) [reply packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts]: We should be consistent about whether we inject these into the endpoints or define them locally. It seems like defining them locally is probably cleaner unless we can imagine a scenario where we'd want to inject a different logger

File: packages/phone-number-privacy/combiner/src/domain/endpoints/sign/action.ts [hunk 13]
```diff
@@ -9,14 +9,15 @@ import {
 } from '@celo/phone-number-privacy-common'
 import { CryptoSession } from '../../../common/crypto-session'
 import { SignAction } from '../../../common/sign'
+import { DomainDiscrepanciesLogger } from '../../services/logDiscrepancies'
 
 export class DomainSignAction extends SignAction<DomainRestrictedSignatureRequest> {
   readonly endpoint: CombinerEndpoint = CombinerEndpoint.DOMAIN_SIGN
   readonly signerEndpoint: SignerEndpoint = getSignerEndpoint(this.endpoint)
+  readonly discrepanyLogger: DomainDiscrepanciesLogger = new DomainDiscrepanciesLogger()
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/domain/endpoints/sign/action.ts]: misspelling

File: packages/phone-number-privacy/combiner/src/pnp/services/logDiscrepancies.ts [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        const { version, performedQueryCount, totalQuota, blockNumber, warnings } = response.res
+        parsedResponses.push({
+          signerUrl: response.url,
+          values: { version, performedQueryCount, totalQuota, blockNumber, warnings },
+        })
+      }
+    })
+    if (parsedResponses.length === 0) {
+      session.logger.warn('No successful signer responses found!')
+      session.warnings.push('No successful signer responses found!')
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/pnp/services/logDiscrepancies.ts]: I'm not sure I see why it would be helpful to send this warning back to the client, could you clarify?

File: packages/phone-number-privacy/combiner/src/pnp/services/logDiscrepancies.ts [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+      return
+    }
+
+    // log all responses if we notice any discrepancies to aid with debugging
+    const first = JSON.stringify(parsedResponses[0].values)
+    for (let i = 1; i < parsedResponses.length; i++) {
+      if (JSON.stringify(parsedResponses[i].values) !== first) {
+        session.logger.warn({ parsedResponses }, 'Discrepancies detected in signer responses')
+        session.warnings.push('Discrepancies detected in signer responses')
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/pnp/services/logDiscrepancies.ts]: Could you create a WarningMessage for this?

File: packages/phone-number-privacy/combiner/src/pnp/services/logDiscrepancies.ts [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+      sortedByTotalQuota[0].values.totalQuota -
+        sortedByTotalQuota[sortedByTotalQuota.length - 1].values.totalQuota >
+      MAX_TOTAL_QUOTA_DISCREPANCY_THRESHOLD
+    ) {
+      session.logger.error(
+        { sortedDescByTotalQuota: sortedByTotalQuota },
+        WarningMessage.INCONSISTENT_SIGNER_QUOTA_MEASUREMENTS
+      )
+      session.warnings.push(WarningMessage.INCONSISTENT_SIGNER_QUOTA_MEASUREMENTS)
+    }
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/pnp/services/logDiscrepancies.ts:91]: Could you please add a block for performedQueryCount like we have for blockNumber and totalQuota?

**Review by alecps — COMMENTED** [review]
Made it through the pnp discrepency logger, still need to go over the domains one

File: packages/phone-number-privacy/combiner/src/config.ts [hunk 17]
```diff
@@ -1,7 +1,7 @@
 import { OdisUtils } from '@celo/identity'
 import { BlockchainConfig, rootLogger, TestUtils, toBool } from '@celo/phone-number-privacy-common'
 import * as functions from 'firebase-functions'
-export const VERSION = process.env.npm_package_version ?? ''
+export const VERSION = process.env.npm_package_version ?? '0.0.0'
```
Comment (alecps) [inline packages/phone-number-privacy/combiner/src/config.ts:4]: No objection but curious why this change was made

**Review by alecps — COMMENTED** [review]
The rest looks good, nice work!

File: packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts (hunk 11 above)
Comment (isabellewei) [reply packages/phone-number-privacy/combiner/src/pnp/endpoints/sign/action.ts]: makes sense! defining them all locally

File: packages/phone-number-privacy/combiner/src/config.ts (hunk 17 above)
Comment (isabellewei) [reply packages/phone-number-privacy/combiner/src/config.ts:4]: it's to match the [default value in the signer](https://github.com/celo-org/celo-monorepo/blob/8ac130841547a86e6f14782c7a1eded39eacd6e1/packages/phone-number-privacy/signer/src/config.ts#L7), as the mismatch was causing the tests to fail

**Review by alecps — APPROVED** [review]
Looks great!! :)
